File: android/app/src/test/kotlin/com/simplephotos/data/album/AlbumSortTest.kt
package com.simplephotos.data.album

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/** A minimal stand-in for the fields the sort reads off a photo. */
private data class Item(val id: String, val takenAt: Long, val name: String)

private fun sort(items: List<Item>, s: AlbumSort): List<String> =
    sortAlbumItems(items, s, { it.takenAt }, { it.name }, { it.id }).map { it.id }

class AlbumSortTest {

    // ── Date ────────────────────────────────────────────────────────────────

    @Test
    fun `date desc puts the newest capture first`() {
        val items = listOf(
            Item("old", 100, "a"),
            Item("new", 300, "b"),
            Item("mid", 200, "c"),
        )
        assertEquals(
            listOf("new", "mid", "old"),
            sort(items, AlbumSort(AlbumSortField.DATE, AlbumSortDir.DESC)),
        )
    }

    @Test
    fun `date asc puts the oldest capture first`() {
        val items = listOf(
            Item("old", 100, "a"),
            Item("new", 300, "b"),
            Item("mid", 200, "c"),
        )
        assertEquals(
            listOf("old", "mid", "new"),
            sort(items, AlbumSort(AlbumSortField.DATE, AlbumSortDir.ASC)),
        )
    }

    @Test
    fun `a missing takenAt sorts as the oldest`() {
        val items = listOf(Item("has", 500, "a"), Item("missing", 0, "b"))
        assertEquals(
            listOf("has", "missing"),
            sort(items, AlbumSort(AlbumSortField.DATE, AlbumSortDir.DESC)),
        )
        assertEquals(
            listOf("missing", "has"),
            sort(items, AlbumSort(AlbumSortField.DATE, AlbumSortDir.ASC)),
        )
    }

    // ── Name (natural / numeric-aware) ────────────────────────────────────────

    @Test
    fun `name asc orders numerically so IMG_2 precedes IMG_10`() {
        val items = listOf(
            Item("x", 0, "IMG_10.jpg"),
            Item("y", 0, "IMG_2.jpg"),
            Item("z", 0, "IMG_1.jpg"),
        )
        assertEquals(
            listOf("z", "y", "x"),
            sort(items, AlbumSort(AlbumSortField.NAME, AlbumSortDir.ASC)),
        )
    }

    @Test
    fun `name sort is case-insensitive`() {
        val items = listOf(Item("x", 0, "banana.jpg"), Item("y", 0, "Apple.jpg"))
        assertEquals(
            listOf("y", "x"),
            sort(items, AlbumSort(AlbumSortField.NAME, AlbumSortDir.ASC)),
        )
    }

    @Test
    fun `name desc reverses the order`() {
        val items = listOf(Item("a", 0, "a.jpg"), Item("b", 0, "b.jpg"), Item("c", 0, "c.jpg"))
        assertEquals(
            listOf("c", "b", "a"),
            sort(items, AlbumSort(AlbumSortField.NAME, AlbumSortDir.DESC)),
        )
    }

    @Test
    fun `naturalCompare handles leading zeros and equal numbers`() {
        // 007 == 7 numerically (leading zeros ignored).
        assertEquals(0, naturalCompare("img007.jpg", "img7.jpg"))
        // 2 < 10 numerically, not lexically.
        assert(naturalCompare("v2", "v10") < 0)
        // Pure text still compares case-folded.
        assert(naturalCompare("apple", "banana") < 0)
    }

    // ── Deterministic ties ────────────────────────────────────────────────────

    @Test
    fun `a date tie breaks on id, stable regardless of input order`() {
        val a = Item("zeta", 100, "z")
        val b = Item("alpha", 100, "a")
        val c = Item("mu", 100, "m")
        val expected = listOf("alpha", "mu", "zeta")
        assertEquals(expected, sort(listOf(a, b, c), AlbumSort(AlbumSortField.DATE, AlbumSortDir.ASC)))
        assertEquals(expected, sort(listOf(c, a, b), AlbumSort(AlbumSortField.DATE, AlbumSortDir.ASC)))
    }

    // ── Toggle logic ──────────────────────────────────────────────────────────

    @Test
    fun `tapping the active field reverses its direction`() {
        val start = AlbumSort(AlbumSortField.DATE, AlbumSortDir.DESC)
        assertEquals(
            AlbumSort(AlbumSortField.DATE, AlbumSortDir.ASC),
            nextSort(start, AlbumSortField.DATE),
        )
    }

    @Test
    fun `tapping a different field switches to its natural direction`() {
        val start = AlbumSort(AlbumSortField.DATE, AlbumSortDir.DESC)
        // Name defaults to ASC (A→Z); Date defaults to DESC (newest first).
        assertEquals(
            AlbumSort(AlbumSortField.NAME, AlbumSortDir.ASC),
            nextSort(start, AlbumSortField.NAME),
        )
        assertEquals(
            AlbumSort(AlbumSortField.DATE, AlbumSortDir.DESC),
            nextSort(AlbumSort(AlbumSortField.NAME, AlbumSortDir.ASC), AlbumSortField.DATE),
        )
    }

    @Test
    fun `defaultDirFor dates desc and names asc`() {
        assertEquals(AlbumSortDir.DESC, defaultDirFor(AlbumSortField.DATE))
        assertEquals(AlbumSortDir.ASC, defaultDirFor(AlbumSortField.NAME))
    }

    // ── Persistence serialization ─────────────────────────────────────────────

    @Test
    fun `serialize round-trips through parse`() {
        for (field in AlbumSortField.entries) {
            for (dir in AlbumSortDir.entries) {
                val s = AlbumSort(field, dir)
                assertEquals(s, parseAlbumSort(s.serialize()))
            }
        }
    }

    @Test
    fun `parse treats missing or malformed values as no choice`() {
        assertNull(parseAlbumSort(null))
        assertNull(parseAlbumSort(""))
        assertNull(parseAlbumSort("garbage"))
        assertNull(parseAlbumSort("size:asc"))
        assertNull(parseAlbumSort("date:sideways"))
        assertNull(parseAlbumSort("date"))
    }
}

File: web/src/gallery/albumSort.test.ts
import { describe, it, expect, beforeEach, beforeAll } from "vitest";
import {
  compareAlbumPhotos,
  sortAlbumPhotos,
  readAlbumSort,
  writeAlbumSort,
  isAlbumSort,
  defaultDirFor,
  DEFAULT_ALBUM_SORT,
  type AlbumSort,
} from "./albumSort";
import type { PhotoWithBurstCount } from "../utils/burstCollapse";

function p(
  blobId: string,
  over: Partial<PhotoWithBurstCount> = {}
): PhotoWithBurstCount {
  return {
    blobId,
    filename: `${blobId}.jpg`,
    takenAt: 0,
    mimeType: "image/jpeg",
    mediaType: "photo",
    width: 1,
    height: 1,
    albumIds: [],
    ...over,
  };
}

const ids = (list: PhotoWithBurstCount[]) => list.map((x) => x.blobId);

// This project's test environment is node (no jsdom), so localStorage is not
// defined. A tiny in-memory stub is enough to exercise the persistence helpers.
beforeAll(() => {
  if (typeof globalThis.localStorage === "undefined") {
    const store = new Map<string, string>();
    globalThis.localStorage = {
      getItem: (k: string) => (store.has(k) ? store.get(k)! : null),
      setItem: (k: string, v: string) => void store.set(k, String(v)),
      removeItem: (k: string) => void store.delete(k),
      clear: () => store.clear(),
      key: (i: number) => [...store.keys()][i] ?? null,
      get length() {
        return store.size;
      },
    } as Storage;
  }
});

describe("compareAlbumPhotos — date", () => {
  it("desc puts the newest capture first", () => {
    const list = [
      p("old", { takenAt: 100 }),
      p("new", { takenAt: 300 }),
      p("mid", { takenAt: 200 }),
    ];
    expect(ids(sortAlbumPhotos(list, { field: "date", dir: "desc" }))).toEqual([
      "new",
      "mid",
      "old",
    ]);
  });

  it("asc puts the oldest capture first", () => {
    const list = [
      p("old", { takenAt: 100 }),
      p("new", { takenAt: 300 }),
      p("mid", { takenAt: 200 }),
    ];
    expect(ids(sortAlbumPhotos(list, { field: "date", dir: "asc" }))).toEqual([
      "old",
      "mid",
      "new",
    ]);
  });

  it("treats a missing takenAt as the oldest (last in desc)", () => {
    const list = [p("has", { takenAt: 500 }), p("missing", { takenAt: 0 })];
    expect(ids(sortAlbumPhotos(list, { field: "date", dir: "desc" }))).toEqual([
      "has",
      "missing",
    ]);
    expect(ids(sortAlbumPhotos(list, { field: "date", dir: "asc" }))).toEqual([
      "missing",
      "has",
    ]);
  });
});

describe("compareAlbumPhotos — name", () => {
  it("orders numerically so IMG_2 precedes IMG_10 (asc)", () => {
    const list = [
      p("x", { filename: "IMG_10.jpg" }),
      p("y", { filename: "IMG_2.jpg" }),
      p("z", { filename: "IMG_1.jpg" }),
    ];
    expect(ids(sortAlbumPhotos(list, { field: "name", dir: "asc" }))).toEqual([
      "z",
      "y",
      "x",
    ]);
  });

  it("is case-insensitive", () => {
    const list = [p("x", { filename: "banana.jpg" }), p("y", { filename: "Apple.jpg" })];
    expect(ids(sortAlbumPhotos(list, { field: "name", dir: "asc" }))).toEqual(["y", "x"]);
  });

  it("desc reverses the name order", () => {
    const list = [
      p("a", { filename: "a.jpg" }),
      p("b", { filename: "b.jpg" }),
      p("c", { filename: "c.jpg" }),
    ];
    expect(ids(sortAlbumPhotos(list, { field: "name", dir: "desc" }))).toEqual([
      "c",
      "b",
      "a",
    ]);
  });
});

describe("compareAlbumPhotos — deterministic ties", () => {
  it("breaks a date tie on blobId so order is stable, not random", () => {
    const list = [
      p("zeta", { takenAt: 100 }),
      p("alpha", { takenAt: 100 }),
      p("mu", { takenAt: 100 }),
    ];
    // Same capture time → deterministic blobId order regardless of input order.
    const asc = ids(sortAlbumPhotos(list, { field: "date", dir: "asc" }));
    expect(asc).toEqual(["alpha", "mu", "zeta"]);
    // A shuffled input yields the identical result — this is the stability
    // JustifiedGrid relies on.
    const shuffled = [list[2], list[0], list[1]];
    expect(ids(sortAlbumPhotos(shuffled, { field: "date", dir: "asc" }))).toEqual(asc);
  });

  it("breaks a name tie on blobId", () => {
    const list = [
      p("b2", { filename: "same.jpg" }),
      p("b1", { filename: "same.jpg" }),
    ];
    expect(ids(sortAlbumPhotos(list, { field: "name", dir: "asc" }))).toEqual(["b1", "b2"]);
  });

  it("does not mutate the input array", () => {
    const list = [p("b", { takenAt: 2 }), p("a", { takenAt: 1 })];
    const before = ids(list);
    sortAlbumPhotos(list, { field: "date", dir: "asc" });
    expect(ids(list)).toEqual(before);
  });
});

describe("compareAlbumPhotos — raw comparator sign", () => {
  it("returns a negative number when a should sort before b", () => {
    const a = p("a", { takenAt: 300 });
    const b = p("b", { takenAt: 100 });
    // desc: newer (300) before older (100).
    expect(compareAlbumPhotos(a, b, { field: "date", dir: "desc" })).toBeLessThan(0);
  });
});

describe("persistence", () => {
  beforeEach(() => localStorage.clear());

  it("round-trips a chosen sort per album id", () => {
    const sort: AlbumSort = { field: "name", dir: "asc" };
    writeAlbumSort("album-7", sort);
    expect(readAlbumSort("album-7")).toEqual(sort);
    // A different album is unaffected.
    expect(readAlbumSort("album-8")).toBeNull();
  });

  it("returns null (intrinsic order) when nothing is stored", () => {
    expect(readAlbumSort("never-set")).toBeNull();
  });

  it("treats a corrupt stored value as no choice", () => {
    localStorage.setItem("albumSort:bad", "{not json");
    expect(readAlbumSort("bad")).toBeNull();
    localStorage.setItem("albumSort:wrong", JSON.stringify({ field: "size", dir: "up" }));
    expect(readAlbumSort("wrong")).toBeNull();
  });
});

describe("helpers", () => {
  it("defaults dates to desc and names to asc", () => {
    expect(defaultDirFor("date")).toBe("desc");
    expect(defaultDirFor("name")).toBe("asc");
  });

  it("DEFAULT_ALBUM_SORT is the historical date-desc order", () => {
    expect(DEFAULT_ALBUM_SORT).toEqual({ field: "date", dir: "desc" });
  });

  it("isAlbumSort rejects malformed values", () => {
    expect(isAlbumSort({ field: "date", dir: "desc" })).toBe(true);
    expect(isAlbumSort({ field: "date" })).toBe(false);
    expect(isAlbumSort(null)).toBe(false);
    expect(isAlbumSort("date")).toBe(false);
  });
});
